static-analysis/autodispose-lint/src/main/kotlin/com/uber/autodispose/lint/AutoDisposeDetector.kt

```diff
@@ -0,0 +1,101 @@
+import org.jetbrains.uast.UCallExpression
+import org.jetbrains.uast.getContainingClass
+import org.jetbrains.uast.getContainingUClass
+import java.io.StringReader
+import java.lang.IllegalStateException
+import java.util.*
+
+internal const val CUSTOM_SCOPE_KEY = "autodispose.classes_with_scope"
+
+/**
+ * Detector which checks if your stream subscriptions are handled by AutoDispose.
+ */
+class AutoDisposeDetector: Detector(), SourceCodeScanner {
+
+  companion object {
+    val ISSUE: Issue = Issue.create(
+        "AutoDisposeUsage",
+        "Always apply an AutoDispose scope before subscribing within defined scoped elements.",
+        "You're subscribing to an observable but not handling it's subscription. This "
+            + "can result in memory leaks. You can avoid memory leaks by appending " +
+            "`.as(autoDisposable(this))` before you subscribe.",
+        Category.CORRECTNESS,
+        10,
+        Severity.ERROR,
+        Implementation(AutoDisposeDetector::class.java, EnumSet.of(Scope.JAVA_FILE)))
+
+    private const val OBSERVABLE = "io.reactivex.Observable"
+    private const val FLOWABLE = "io.reactivex.Flowable"
+    private const val PARALLEL_FLOWABLE = "io.reactivex.parallel.ParallelFlowable"
+    private const val SINGLE = "io.reactivex.Single"
+    private const val MAYBE = "io.reactivex.Maybe"
+    private const val COMPLETABLE = "io.reactivex.Completable"
+  }
+
+  private val defaultScopes = mutableSetOf("androidx.lifecycle.LifecycleOwner",
+      "com.uber.autodispose.ScopeProvider",
+      "com.uber.autodispose.lifecycle.LifecycleScopeProvider",
+      "android.app.Activity",
+      "android.app.Fragment")
+
+  override fun beforeCheckRootProject(context: Context) {
+    val props = Properties()
+    context.project.propertyFiles.find { it.name == "local.properties" }?.apply {
+      val content = StringReader(context.client.readFile(this).toString())
+      props.load(content)
+      val scopes = props.getProperty(CUSTOM_SCOPE_KEY).split(",").map { it.trim() }
+      defaultScopes.addAll(scopes)
+    }
+  }
+
+  override fun getApplicableMethodNames(): List<String> = listOf("subscribe")
+
+  override fun visitMethod(context: JavaContext, node: UCallExpression, method: PsiMethod) {
+    val evaluator = context.evaluator
+
+    if (isReactiveType(evaluator, method) && isInScope(evaluator, node.getContainingUClass())) {
+      context.report(ISSUE, node, context.getLocation(node), "Always apply an AutoDispose " +
+          "scope before subscribing within defined scoped elements.")
+    }
+  }
+
+  private fun isInScope(evaluator: JavaEvaluator, psiClass: PsiClass?): Boolean {
+    psiClass?.let { callingClass ->
+      return defaultScopes.any {
+        evaluator.inheritsFrom(callingClass, it, false)
+      }
+    }
+    return false
+  }
+
+  private fun isReactiveType(evaluator: JavaEvaluator, method: PsiMethod): Boolean {
+    return evaluator.isMemberInClass(method, OBSERVABLE) ||
```

Comment:
Maybe we can put all the types in a set and just do a `rxTypes.any { evaluator.isMemeberInClass(method, it) }`?